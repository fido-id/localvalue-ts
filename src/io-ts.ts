import { pipe } from "fp-ts/lib/function"
import * as t from "io-ts"
import * as E from "fp-ts/Either"
import { Json, JsonFromString } from "io-ts-types"
import * as LV from "./LocalValue"
import { Codec } from "./Codec"

const adaptIoTsCodec = <A, B>(C: t.Type<B, A>): Codec<t.Errors, A, B> => {
  return {
    encode: C.encode,
    decode: (u: unknown) => LV.fromEither(C.decode(u)),
  }
}

export const fromIoTsCodec = <A, B extends Json>(C: t.Type<A, B>) => {
  const stringCodec = new t.Type<A, string>(
    C.name,
    C.is,
    (u, c) => {
      return pipe(
        t.string.validate(u, c),
        E.chain((jsonString) => JsonFromString.validate(jsonString, c)),
        E.chain((json) => C.validate(json, c)),
      )
    },
    (v) => {
      return pipe(v, C.encode, JsonFromString.encode)
    },
  )

  return adaptIoTsCodec(stringCodec)
}
