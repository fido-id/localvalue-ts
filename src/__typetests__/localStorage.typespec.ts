import * as t from "io-ts"
import { DateFromISOString } from "io-ts-types"
import { fromIoTsCodec } from "../io-ts"
import {
  createLocalStorage,
  getLocalValue,
  removeLocalValue,
  setLocalValue,
} from "../localStorage"

const ShapeCodec = t.type({ s: t.string, d: DateFromISOString })
type ShapeCodec = t.TypeOf<typeof ShapeCodec>

const CorrectCodec = fromIoTsCodec(ShapeCodec)

export const defaultShape: ShapeCodec = {
  s: "foo",
  d: new Date(1600732800 * 1000),
}

// @dts-jest:pass:snap It works with string encoding
getLocalValue("sape", CorrectCodec)

// @dts-jest:fail:snap It doesn't work with non-string encoding
getLocalValue("shape", ShapeCodec)

// @dts-jest:pass:snap You can pass a valid set of options
getLocalValue("sape", CorrectCodec, {
  defaultValue: { s: "foo", d: new Date() },
  useMemorySore: false,
})

getLocalValue("sape", CorrectCodec, {
  // @dts-jest:fail:snap You cannot pass an invalid set of options
  defaultValue: { foo: 123 },
  useMemorySore: false,
})

// @dts-jest:pass:snap It works with string encoding
setLocalValue("sape", CorrectCodec, defaultShape)

// @dts-jest:fail:snap It doesn't work with non-string encoding
setLocalValue("shape", ShapeCodec, defaultShape)

// @dts-jest:pass:snap You can pass a valid set of options
setLocalValue("shape", CorrectCodec, defaultShape, {
  defaultValue: { s: "foo", d: new Date() },
  useMemorySore: false,
})

setLocalValue("shape", CorrectCodec, defaultShape, {
  // @dts-jest:fail:snap You cannot pass an invalid set of options
  defaultValue: { foo: 123 },
  useMemorySore: false,
})

// @dts-jest:pass:snap It works with any string
removeLocalValue("shape")

// @dts-jest:pass:snap You can pass a valid set of options
removeLocalValue("shape", {
  defaultValue: { s: "foo", d: new Date() },
  useMemorySore: false,
})

removeLocalValue("shape", {
  // @dts-jest:fail:snap You cannot pass an invalid set of options
  useMemorySore: "false",
})

const store = createLocalStorage({
  // @dts-jest:pass:snap It works with string encoding
  foo: CorrectCodec,
})

createLocalStorage({
  // @dts-jest:fail:snap It doesn't work with non-string encoding
  foo: ShapeCodec,
})

const storeWithOptions = createLocalStorage(
  {
    foo: CorrectCodec,
  },
  // @dts-jest:pass:snap You can pass a valid set of options to store
  { useMemorySore: true, defaultValues: { foo: defaultShape } },
)

createLocalStorage(
  {
    foo: CorrectCodec,
  },
  // @dts-jest:fail:snap You cannot pass an invalid set of options to store
  { useMemorySore: true, defaultValues: { fo: defaultShape } },
)

// @dts-jest:pass:snap store returns the correct type encoding
store

// @dts-jest:pass:snap storeWithOptions returns the correct type encoding
storeWithOptions
