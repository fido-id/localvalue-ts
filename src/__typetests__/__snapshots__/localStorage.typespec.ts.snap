// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`It doesn't work with non-string encoding (type) should match snapshot 1`] = `
"Argument of type 'TypeC<{ s: StringC; d: DateFromISOStringC; }>' is not assignable to parameter of type 'Codec<unknown, string, { s: string; d: Date; }>'.
  The types returned by 'encode(...)' are incompatible between these types.
    Type '{ s: string; d: string; }' is not assignable to type 'string'."
`;

exports[`It doesn't work with non-string encoding (type) should match snapshot 2`] = `"Argument of type 'TypeC<{ s: StringC; d: DateFromISOStringC; }>' is not assignable to parameter of type 'Codec<unknown, string, { s: string; d: Date; }>'."`;

exports[`It doesn't work with non-string encoding (type) should match snapshot 3`] = `
"Type 'TypeC<{ s: StringC; d: DateFromISOStringC; }>' is not assignable to type 'Codec<any, string, any>'.
  The types returned by 'encode(...)' are incompatible between these types.
    Type '{ s: string; d: string; }' is not assignable to type 'string'."
`;

exports[`It works with any string (type) should match snapshot 1`] = `"void"`;

exports[`It works with string encoding (type) should match snapshot 1`] = `"LocalValue<Errors, { s: string; d: Date; }>"`;

exports[`It works with string encoding (type) should match snapshot 2`] = `"void"`;

exports[`It works with string encoding (type) should match snapshot 3`] = `"Codec<Errors, string, { s: string; d: Date; }>"`;

exports[`You can pass a valid set of options (type) should match snapshot 1`] = `"LocalValue<Errors, { s: string; d: Date; }>"`;

exports[`You can pass a valid set of options (type) should match snapshot 2`] = `"void"`;

exports[`You can pass a valid set of options (type) should match snapshot 3`] = `"void"`;

exports[`You can pass a valid set of options to store (type) should match snapshot 1`] = `"any"`;

exports[`You cannot pass an invalid set of options (type) should match snapshot 1`] = `
"Type '{ foo: number; }' is not assignable to type '{ s: string; d: Date; }'.
  Object literal may only specify known properties, and 'foo' does not exist in type '{ s: string; d: Date; }'."
`;

exports[`You cannot pass an invalid set of options (type) should match snapshot 2`] = `
"Type '{ foo: number; }' is not assignable to type '{ s: string; d: Date; }'.
  Object literal may only specify known properties, and 'foo' does not exist in type '{ s: string; d: Date; }'."
`;

exports[`You cannot pass an invalid set of options (type) should match snapshot 3`] = `"Type 'string' is not assignable to type 'boolean'."`;

exports[`You cannot pass an invalid set of options to store (type) should match snapshot 1`] = `
"Type '{ fo: { s: string; d: Date; }; }' is not assignable to type '{ foo: { s: string; d: Date; }; }'.
  Object literal may only specify known properties, and 'fo' does not exist in type '{ foo: { s: string; d: Date; }; }'."
`;

exports[`store returns the correct type encoding (type) should match snapshot 1`] = `"{ foo: { getValue: () => LocalValue<Errors, { s: string; d: Date; }>; setValue: (v: { s: string; d: Date; }) => void; removeValue: () => void; }; }"`;

exports[`storeWithOptions returns the correct type encoding (type) should match snapshot 1`] = `"{ foo: { getValue: () => LocalValue<Errors, { s: string; d: Date; }>; setValue: (v: { s: string; d: Date; }) => void; removeValue: () => void; }; }"`;
